src/sys/pthread.rs

```diff
@@ -34,10 +41,285 @@ pub fn pthread_kill<T>(thread: Pthread, signal: T) -> Result<()>
+            uninit.assume_init()
+        };
+        Ok(Protocol::from(init))
+    }
+    /// Wraps [`libc::pthread_mutexattr_setprotocol`].
+    pub fn set_protocol(&mut self, protocol: Protocol) -> Result<()> {
+        unsafe {
+            Errno::result(libc::pthread_mutexattr_setprotocol(&mut self.0,protocol as i32)).map(drop)
+        }
+    }
+}
+
+#[cfg(target_os = "linux")]
+impl Default for MutexAttr {
+    fn default() -> Self {
+        let mutex_attr = Self::new().unwrap();
+        debug_assert_eq!(mutex_attr.get_shared(),Ok(true));
+        mutex_attr
+    }
+}
+#[cfg(target_os = "linux")]
+impl std::ops::Drop for MutexAttr {
+    /// Wraps [`libc::pthread_mutexattr_destroy`].
+    fn drop(&mut self) {
+        unsafe {
+            Errno::result(libc::pthread_mutexattr_destroy(&mut self.0)).unwrap();
+        }
+    }
+}
+
+/// Mutex.
+/// ```
+/// # use std::{
+/// #   sync::Arc,
+/// #   time::{Instant, Duration},
+/// #   thread::{sleep, spawn},
+/// #   mem::size_of,
+/// #   num::NonZeroUsize,
+/// #   os::unix::io::OwnedFd
+/// # };
+/// # use nix::{
+/// #   sys::{pthread::{Mutex, MutexAttr}, mman::{mmap, MapFlags, ProtFlags}},
+/// #   unistd::{fork,ForkResult},
+/// # };
+/// const TIMEOUT: Duration = Duration::from_millis(500);
+/// const DELTA: Duration = Duration::from_millis(100);
+/// # fn main() -> nix::Result<()> {
+/// let mutex = Mutex::default();
+/// 
+/// // The mutex is initialized unlocked, so an attempt to unlock it should
+/// // return immediately.
+/// assert_eq!(mutex.unlock(), Ok(()));
+/// // The mutex is unlocked, so `try_lock` will lock.
+/// assert_eq!(mutex.try_lock(), Ok(true));
+/// // Unlock the mutex.
+/// assert_eq!(mutex.unlock(), Ok(()));
+/// // The mutex is unlocked, so `lock` will lock and exit immediately.
+/// assert_eq!(mutex.lock(), Ok(()));
+/// // Unlock the mutex.
+/// assert_eq!(mutex.unlock(), Ok(()));
+/// 
+/// // Test across threads
+/// // -------------------------------------------------------------------------
+/// 
+/// let mutex = Arc::new(mutex);
+/// let mutex_clone = mutex.clone();
+/// let instant = Instant::now();
+/// spawn(move || {
+///     assert_eq!(mutex_clone.lock(), Ok(()));
+///     sleep(TIMEOUT);
+///     assert_eq!(mutex_clone.unlock(), Ok(()));
+/// });
+/// sleep(DELTA);
+/// assert_eq!(mutex.lock(), Ok(()));
+/// assert!(instant.elapsed() > TIMEOUT && instant.elapsed() < TIMEOUT + DELTA);
+/// 
+/// // Test across processes
+/// // -------------------------------------------------------------------------
+///
+/// let shared_memory = unsafe { mmap::<OwnedFd>(
+///     None,
+///     NonZeroUsize::new_unchecked(size_of::<Mutex>()),
+///     ProtFlags::PROT_WRITE | ProtFlags::PROT_READ,
+///     MapFlags::MAP_SHARED | MapFlags::MAP_ANONYMOUS,
+///     None,
+///     0
+/// )? };
+/// let mutex_ptr = shared_memory.cast::<Mutex>();
+/// let mutex = unsafe { &*mutex_ptr };
```

Comment:
This line creates a `Mutex` with uninitialized memory, right?  I"m not sure that's ok.  Safer to initialize it via passing a raw pointer to C.  Like with a `fn init(*mut Mutex, attr: Option<MutexAttr>) -> Result<Self>` method.  Then you can trust that the memory will always be initialized before you access it from Rust code.  You'll need `repr(transparent` for that to be guaranteed to work.

Reply:
Changed.